genericfunc.go

```diff
@@ -0,0 +1,106 @@
+package linq
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// GenericType represents a any reflect.Type
+type GenericType int
+
+var genericType = reflect.TypeOf(new(GenericType)).Elem()
+
+// FunctionCache keeps in cached the GenericFunc reflections objects
+type FunctionCache struct {
+	FnValue  reflect.Value
+	FnType   reflect.Type
+	TypesIn  []reflect.Type
+	TypesOut []reflect.Type
+}
+
+// GenericFunc is a type used to validate and call dynamic functions
+type GenericFunc struct {
+	Cache *FunctionCache
+}
+
+// Call calls a dynamic function
+func (d *GenericFunc) Call(params ...interface{}) []interface{} {
+	paramsIn := make([]reflect.Value, len(params))
+	for i, param := range params {
+		paramsIn[i] = reflect.ValueOf(param)
+	}
+	paramsOut := d.Cache.FnValue.Call(paramsIn)
+	interfaceOut := make([]interface{}, len(paramsOut))
```

Comment:
probably do not need that allocation, because there are no predicates that return multiple results, and this method is performance critical.